---
title: "Untitled"
output: html_document
date: "2024-11-13"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Question 1.1

Let the proportion of allele A in the population be $P_0(A)=p$

Let the proportion of allele a in the population be $P_0(a)=q$

The probability of an offspring AA in the first generation should be $P_1(AA) = p \cdot p = p^2$

The probability of an offspring aa in the first generation should be $P_1(aa) = q \cdot q = q^2$

The probability of an offspring Aa in the first generation should be $P_1(Aa) +P_1(aA)= p \cdot q + q\cdot p=2pq$

The proportion of A in the first generation is therefore $P_1(A)= P_1(AA)+\frac{1}{2} P_1(Aa)=p^2+pq=p(p+q)=p$

The proportion of A in the first generation is therefore $P_1(a)= P_1(aa)+\frac{1}{2} P_1(Aa)=q^2+pq=p(p+q)=q$

Since $P_0(A) = P_1(A) = p$ and $P_0(a)=P_1(a) = q$ , the allelle proportion are the same between generation 0 and generation 1. This will also be true for every following generation, hence Hardy-Weinberg equlibrium is reached.

## Question 1.2

```{r}
MM <- 357
MN <- 485
NN <- 158

p <- (MM*2 + MN) / 2000
q <- 1-p

chisq.test(x = c(MM, MN, NN), p = c(p**2, 2*p*q, q**2))
```

Hence, it appears as though the population in a state of Hardy-Weinberg equilibrium. Chi-square-goodness-of-fit shows that we cannot reject the null-hypothesis that the population allelle distribution is as expected.

## Question 2.1

The organism in question is a Branchipus schaefferi, more commonly known as a fairy shrimp. The protein product of the CDS is named **cytochrome c oxidase subunit I**

## Question 2.2

```{r}

# Protein Sequences

Transeq <- "LLGDDQLYNVIVTAHAFVMIFFMVMPILIGGFGNWLVPLMLGAPDMAFPRLNNLSFWMLPPSLTLLVASSMVESGVGTGWTVYPPLSAAIAHAGPSVDLAIFSLHLAGISSILGAVNFITTVINMRPHSMSLDRMPLFAWAVVITAVLLLLSLPVLAX"
GenBank <- "LLGDDQLYNVIVTAHAFVMIFFMVMPILIGGFGNWLVPLMLGAPDMAFPRLNNLSFWMLPPSLTLLVASSMVESGVGTGWTVYPPLSAAIAHAGPSVDLAIFSLHLAGISSILGAVNFITTVINMRPHSMSLDRMPLFAWAVVITAVLLLLSLPVLA"


# DNA Sequences

GenBankOrigin <- "TCTCCTAGGAGATGACCAACTTTATAACGTCATTGTTACTGCTCACGCTTTTGTAATGATTTTCTTCATAGTTATACCAATCCTTATTGGAGGATTTGGAAATTGATTAGTCCCTTTAATACTAGGGGCTCCTGATATGGCTTTCCCCCGACTAAATAACTTAAGCTTTTGAATACTTCCTCCCTCATTAACTCTTCTAGTGGCCAGCTCAATGGTAGAAAGAGGGGTAGGAACAGGATGAACAGTTTATCCACCCCTATCTGCTGCTATTGCCCATGCTGGTCCTTCTGTTGATTTAGCAATCTTTTCACTTCACCTCGCAGGGATCTCTTCAATTTTAGGAGCTGTAAATTTCATTACAACTGTAATTAATATACGGCCTCATTCCATAAGATTAGACCGAATACCTTTATTTGCATGAGCGGTAGTTATCACAGCAGTTCTTCTTCTCCTTTCTCTCCCAGTATTAGCAG"

Backtranseq <- "CTGCTGGGCGATGATCAGCTGTACAACGTGATCGTGACCGCCCACGCCTTCGTGATGATCTTCTTCATGGTGATGCCCATCCTGATCGGCGGCTTCGGCAACTGGCTGGTGCCCCTGATGCTGGGCGCCCCCGATATGGCCTTCCCCCGCCTGAACAACCTGAGCTTCTGGATGCTGCCCCCCAGCCTGACCCTGCTGGTGGCCAGCAGCATGGTGGAGAGCGGCGTGGGCACCGGCTGGACCGTGTACCCCCCCCTGAGCGCCGCCATCGCCCACGCCGGCCCCAGCGTGGATCTGGCCATCTTCAGCCTGCACCTGGCCGGCATCAGCAGCATCCTGGGCGCCGTGAACTTCATCACCACCGTGATCAACATGCGCCCCCACAGCATGAGCCTGGATCGCATGCCCCTGTTCGCCTGGGCCGTGGTGATCACCGCCGTGCTGCTGCTGCTGAGCCTGCCCGTGCTGGCCNNN"

BacktranseqRevComp <- "NNNGGCCAGCACGGGCAGGCTCAGCAGCAGCAGCACGGCGGTGATCACCACGGCCCAGGCGAACAGGGGCATGCGATCCAGGCTCATGCTGTGGGGGCGCATGTTGATCACGGTGGTGATGAAGTTCACGGCGCCCAGGATGCTGCTGATGCCGGCCAGGTGCAGGCTGAAGATGGCCAGATCCACGCTGGGGCCGGCGTGGGCGATGGCGGCGCTCAGGGGGGGGTACACGGTCCAGCCGGTGCCCACGCCGCTCTCCACCATGCTGCTGGCCACCAGCAGGGTCAGGCTGGGGGGCAGCATCCAGAAGCTCAGGTTGTTCAGGCGGGGGAAGGCCATATCGGGGGCGCCCAGCATCAGGGGCACCAGCCAGTTGCCGAAGCCGCCGATCAGGATGGGCATCACCATGAAGAAGATCATCACGAAGGCGTGGGCGGTCACGATCACGTTGTACAGCTGATCATCGCCCAGCAG"






```

## Question 2.3

As displayed in the code chunks where the translation results are stored, a similar result is obtained when translating the nucleotides into protein the difference is that the translated protein ends with an "X".

When translating back, we get a very different result. Reversing-complementing the sequence does not seem to result in a similar sequence either.

```{r}
Backtranambig <- "YTNYTNGGNGAYGAYCARYTNTAYAAYGTNATYGTNACNGCNCAYGCNTTYGTNATRATYTTYTTYATRGTNATRCCNATYYTNATYGGNGGNTTYGGNAAYTGRYTNGTNCCNYTNATRYTNGGNGCNCCNGAYATRGCNTTYCCNCGNYTNAAYAAYYTNWSNTTYTGRATRYTNCCNCCNWSNYTNACNYTNYTNGTNGCNWSNWSNATRGTNGARWSNGGNGTNGGNACNGGNTGRACNGTNTAYCCNCCNYTNWSNGCNGCNATYGCNCAYGCNGGNCCNWSNGTNGAYYTNGCNATYTTYWSNYTNCAYYTNGCNGGNATYWSNWSNATYYTNGGNGCNGTNAAYTTYATYACNACNGTNATYAAYATRCGNCCNCAYWSNATRWSNYTNGAYCGNATRCCNYTNTTYGCNTGRGCNGTNGTNATYACNGCNGTNYTNYTNYTNYTNWSNYTNCCNGTNYTNGCNNNN"

# Codon Table 5
translate <- list()
translate[["A"]] <- "A"
translate[["T"]] <- "T"
translate[["G"]] <- "G"
translate[["C"]] <- "C"
translate[["Y"]] <- c("C","T")
translate[["R"]] <- c("A","G")
translate[["S"]] <- c("G","C")
translate[["W"]] <- c("A","T")
translate[["K"]] <- c("T","G")
translate[["M"]] <- c("A","C")
translate[["B"]] <- c("C","G","T")
translate[["D"]] <- c("A","G","T")
translate[["H"]] <- c("A","C","T")
translate[["V"]] <- c("A","C","G")
translate[["N"]] <- c("A","C","G","T")

df <- data.frame(GenBankOrigin = NA, Backtranambig = NA, Candidates = NA, MatchLetter = NA, Match = NA)

for(i in 1:(nchar(GenBankOrigin)-1)){
  df[i,1] <- strsplit(GenBankOrigin, "")[[1]][i]
  df$Backtranambig[i] <- strsplit(Backtranambig, "")[[1]][i]
  df$Candidates[i] <- paste0(translate[[df$Backtranambig[i]]], collapse = ",")
  df$Match[i] <- df$GenBankOrigin[i] %in%  translate[[df$Backtranambig[i]]]
  if(df$Match[i]) df$MatchLetter[i] <- df$GenBankOrigin[i] 
}

head(df,25)
```

Backtranambig gives as a sequence to test for compability when ambiguity is suspected. Ambiguous translations can occur because several codons can translate into the same amino acid. The sequences displayed above does not seem to be compatible as the bases does not match.

This is likely caused by the fact that GenBankOrigin is set to start at the base of index 2, not 1. Below is the code rewritten with a shift +1 to the index.

```{r}
df <- data.frame(GenBankOrigin = NA, Backtranambig = NA, Candidates = NA, MatchLetter = NA, Match = NA)

for(i in 1:(nchar(GenBankOrigin)-1)){
  df[i,1] <- strsplit(GenBankOrigin, "")[[1]][i+1] # +1 here
  df$Backtranambig[i] <- strsplit(Backtranambig, "")[[1]][i]
  df$Candidates[i] <- paste0(translate[[df$Backtranambig[i]]], collapse = ",")
  df$Match[i] <- df$GenBankOrigin[i] %in%  translate[[df$Backtranambig[i]]]
  if(df$Match[i]) df$MatchLetter[i] <- df$GenBankOrigin[i] 
}

head(df,25)
```

Now, we can see that our sequences are compatible. But up until now, only the GenBank-origin sequence has been tested for compability with itself.

Below, the backtranslated (Backtranseq) should be checked for compability with the GenBank-origin.

```{r}
df <- data.frame(Backtranseq = NA, Backtranambig = NA, Candidates = NA, MatchLetter = NA, Match = NA)

for(i in 1:(nchar(GenBankOrigin)-1)){
  df[i,1] <- strsplit(Backtranseq, "")[[1]][i]
  df$Backtranambig[i] <- strsplit(Backtranambig, "")[[1]][i]
  df$Candidates[i] <- paste0(translate[[df$Backtranambig[i]]], collapse = ",")
  df$Match[i] <- df$Backtranseq[i] %in%  translate[[df$Backtranambig[i]]]
  if(df$Match[i]) df$MatchLetter[i] <- df$Backtranseq[i] 
}

head(df,200)
```

This shows that the Backtranslated sequences does seem inaccurate because of ambiguous translations, although it is proven compatible. For example, the first 9 (10 because of +1 shift) bases of the GenBank-sequence (*T*) CTC CTA GGA translates to the same amino acids as the backtranslated sequence CTG CTG GGC.

The translational ambiguousess is one-way, hence the translation from DNA to Protein is seemingly more accurate than Protein to DNA. It is also important to translate using correct index shift and correct codon table.
