---
title: "Bioinformatics Lab 2"
author: "Simon Jorstedt"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r}
# Dependencies and data
library(ape)
library(dplyr)

lizards_accession_numbers <- c("JF806202", "HM161150", "FJ356743", "JF806205", 
                               "JQ073190", "GU457971", "FJ356741", "JF806207",
                               "JF806210", "AY662592", "AY662591", "FJ356748",       
                               "JN112660", "AY662594", "JN112661", "HQ876437", 
                               "HQ876434", "AY662590", "FJ356740", "JF806214", 
                               "JQ073188", "FJ356749", "JQ073189", "JF806216", 
                               "AY662598", "JN112653", "JF806204", "FJ356747", 
                               "FJ356744", "HQ876440", "JN112651", "JF806215",
                               "JF806209")

lizards_sequences<-ape::read.GenBank(lizards_accession_numbers)
print(lizards_sequences)
ape::write.dna(lizards_sequences,
               file = "lizard_seqs.fasta", 
               format = "fasta", 
               append = FALSE, 
               nbcol = 6,
               colsep = " ", 
               colw = 10)
```

# Question 1
## Q 1.1
*Simulate an artificial DNA sequence dataset. It should contain 33 sequences. The lengths of the sequences should be the same as in the lizard dataset, i.e. for each real sequence simulate an arti cial one. The simulation rule is as follows, each nucleotide is to be independently and randomly drawn from the distribution given by the base composition (frequencies) in the true lizard sequences. Save your dataset in a fasta format le. Remember to give unique names to your sequences. Report on the base composition in your simulated data.*

```{r}
sim_sequences <- list()

# Simulate sequences
for (i in 1:33){
  # Concatenate a vector like c("a", ..., "g")
  sim_sequences[[i]] <- paste(
    # Sample 
    sample(x = c("a", "c", "g", "t"),
           size = lizards_sequences[[i]] %>% length(),
           replace = TRUE,
           prob = base.freq(lizards_sequences[i])),
    collapse = "")
}

# Rename, convert to DNAbin, and save the sequences to a fasta file
names(sim_sequences) <- lizards_accession_numbers
sim_sequences <- as.DNAbin(sapply(sim_sequences, strsplit, split = ""))

ape::write.dna(sim_sequences,
               file ="simulated_sequences.fasta",
               format = "fasta")
```

The base composition of the respective simulated sequences will be roughly the same as the true base compositions, due to the law of large numbers. Example:

```{r}
# Base frequences of the first real sequence
base.freq(lizards_sequences[1])

# Base frequences of the first simulated sequence
base.freq(sim_sequences[1])
```

# Question 2
## Q 2.1


































