---
title: "Bioinformatics Lab 2"
author: "Simon Jorstedt"
date: "`r Sys.Date()`"
output: pdf_document
---

```{r}
# Dependencies and data
library(ape)
library(dplyr)
library(msa)
library(seqinr)

lizards_accession_numbers <- c("JF806202", "HM161150", "FJ356743", "JF806205", 
                               "JQ073190", "GU457971", "FJ356741", "JF806207",
                               "JF806210", "AY662592", "AY662591", "FJ356748",       
                               "JN112660", "AY662594", "JN112661", "HQ876437", 
                               "HQ876434", "AY662590", "FJ356740", "JF806214", 
                               "JQ073188", "FJ356749", "JQ073189", "JF806216", 
                               "AY662598", "JN112653", "JF806204", "FJ356747", 
                               "FJ356744", "HQ876440", "JN112651", "JF806215",
                               "JF806209")

lizards_sequences<-ape::read.GenBank(lizards_accession_numbers)
print(lizards_sequences)
ape::write.dna(lizards_sequences,
               file = "lizard_seqs.fasta", 
               format = "fasta", 
               append = FALSE, 
               nbcol = 6,
               colsep = " ", 
               colw = 10)
```

# Question 1
## Q 1.1
*Simulate an artificial DNA sequence dataset. It should contain 33 sequences. The lengths of the sequences should be the same as in the lizard dataset, i.e. for each real sequence simulate an arti cial one. The simulation rule is as follows, each nucleotide is to be independently and randomly drawn from the distribution given by the base composition (frequencies) in the true lizard sequences. Save your dataset in a fasta format le. Remember to give unique names to your sequences. Report on the base composition in your simulated data.*

```{r}
set.seed(8182746)

sim_sequences <- list()

# Simulate sequences
for (i in 1:33){
  # Concatenate a vector like c("a", ..., "g")
  sim_sequences[[i]] <- paste(
    # Sample 
    sample(x = c("a", "c", "g", "t"),
           size = lizards_sequences[[i]] %>% length(),
           replace = TRUE,
           prob = base.freq(lizards_sequences[i])),
    collapse = "")
}

# Rename, convert to DNAbin, and save the sequences to a fasta file
names(sim_sequences) <- lizards_accession_numbers
sim_sequences <- as.DNAbin(sapply(sim_sequences, strsplit, split = ""))

ape::write.dna(sim_sequences,
               file ="simulated_sequences.fasta",
               format = "fasta")
```

The base composition of the respective simulated sequences will be roughly the same as the true base compositions, due to the law of large numbers. Example:

```{r}
# Base frequences of the first real sequence
base.freq(lizards_sequences[1])

# Base frequences of the first simulated sequence
base.freq(sim_sequences[1])
```

# Question 2
## Q 2.1
*Report some basic statistics on each sequence dataset: individual base composition, GC content, CpG, ApT content. Also translate your sequences into protein sequences (see Lab 1) and report on the amino acid composition. In your simulated sequences, how many times did you observe a stop codon inside your sequence? Does this occur in your true sequences? Comment.*

```{r}
# Create empty dataframe
df_base_stats <- data.frame(Accession=character(),
                            GC_cont=numeric(),
                            CpG_cont=numeric(),
                            ApT_cont=numeric())


for(i in 1:33){
  # GC content
  gc_cont <- base.freq(lizards_sequences[1])[c(2,3)] %>% sum()
  
  # CpG and ApT contents
  dna_string <- as.character(lizards_sequences)[[i]] %>% paste(collapse = "")
  CpG_cont <- gregexpr(pattern="cg", text=dna_string)[[1]] %>% length() %>% {2*./nchar(dna_string)}
  ApT_cont <- gregexpr(pattern="at", text=dna_string)[[1]] %>% length() %>% {2*./nchar(dna_string)}
  
  # Store values
  new_row <- data.frame(Accession=lizards_accession_numbers[i],
                        GC_cont=gc_cont,
                        CpG_cont=CpG_cont,
                        ApT_cont=ApT_cont)
  
  df_base_stats <- rbind(df_base_stats, new_row)
}

df_base_stats
```

We will now translate the sequences to proteins.

```{r, warning=F}
# Translate to proteins
# Get codon table and Open Reading Frame (ORF)
cod_orf_info <- read.delim("codon_and_orf.txt", header = TRUE, sep = ",")

lizard_proteins <- list()
sim_proteins <- list()
for(i in 1:33){
  # Translate true lizard sequences
  true_protein <- ape::trans(lizards_sequences[[i]],
                             codonstart = cod_orf_info["codon_start"][[1]][i],
                             code = cod_orf_info["transl_table"][[1]][i])
  lizard_proteins[[i]] <- true_protein
  
  # Translate simulated sequences
  sim_prot <- ape::trans(sim_sequences[[i]],
                         codonstart = cod_orf_info["codon_start"][[1]][i],
                         code = cod_orf_info["transl_table"][[1]][i])
  sim_proteins[[i]] <- sim_prot
}
```

Now we can determine the composition of amino acids in the proteins.

```{r}
# Report amino acid compositions
for(i in 1:33){
  true_composition <- as.character(lizard_proteins[[1]]) %>%
    table(dnn = paste("Composition of Amino Acids in",
                      lizards_accession_numbers[i]))
  
  sim_compositions <- as.character(sim_proteins[[1]]) %>%
    table(dnn = paste("Composition of Amino Acids in",
                      lizards_accession_numbers[i], "(simulated)"))
  
  print(true_composition)
  print(sim_compositions)
  cat("\n")
}
```

Now we will count how many times the stop codons (taa, tag, tga) occur in the respective sequences. Because the simulated sequences are random, the stop codons will occur frequently.
```{r, message=FALSE}
# Positions of stop codons in simulated sequence (example)

sim_seq_n_stopcodons <- c()
liz_seq_n_stopcodons <- c()
for(i in 1:33){
  ### Get number of stop codons in *simulated* sequences
  n_taa <- as.character(sim_sequences[[i]] %>%
                          paste(collapse="")) %>%
    gregexpr(pattern="taa", text=.) %>%
    unlist() %>% length()
  
  n_tag <- as.character(sim_sequences[[i]] %>%
                          paste(collapse="")) %>%
    gregexpr(pattern="tag", text=.) %>%
    unlist() %>% length()
  
  n_tga <- as.character(sim_sequences[[i]] %>%
                          paste(collapse="")) %>%
    gregexpr(pattern="tga", text=.) %>%
    unlist() %>% length()
  
  sim_seq_n_stopcodons <- c(sim_seq_n_stopcodons, n_taa+n_tag+n_tga)
  
  
  
  ### Get number of stop codons in *true* sequences
  n_taa <- as.character(lizards_sequences)[[i]] %>%
                          paste(collapse="") %>%
    gregexpr(pattern="taa", text=.) %>%
    unlist() %>% length()
  
  n_tag <- as.character(lizards_sequences)[[i]] %>%
                          paste(collapse="") %>%
    gregexpr(pattern="tag", text=.) %>%
    unlist() %>% length()
  
  n_tga <- as.character(lizards_sequences)[[i]] %>%
                          paste(collapse="") %>%
    gregexpr(pattern="tga", text=.) %>%
    unlist() %>% length()
  
  liz_seq_n_stopcodons <- c(liz_seq_n_stopcodons, n_taa+n_tag+n_tga)
}
print(sim_seq_n_stopcodons)
print(liz_seq_n_stopcodons)

#cat("Number of stop codons in simulated sequences:\n", sim_seq_n_stopcodons, "\n")
#cat("Number of stop codons in true sequences:\n", liz_seq_n_stopcodons, "\n")

# Positions of stop codons in true sequenes
#as.character(lizards_sequences)[[1]] %>% paste(collapse = "") %>%
#  gregexpr(pattern="taa", text=.) %>% unlist()
```

```{r}
# Write lizard proteins in a fasta
ape::write.dna(lizard_proteins,
               file = "lizard_proteins.fasta",
               format = "fasta", 
               append = FALSE, 
               nbcol = 6,
               colsep = "", 
               colw = 10)

# True lizard protein alignment heatmap
readAAStringSet(filepath = "lizard_proteins.fasta") %>%
  msa() %>%
  msaConvert(., type="seqinr::alignment") %>%
  dist.alignment(., "identity") %>%
  as.matrix(.) %>%
  heatmap()


# Write simulated proteins in a fasta
ape::write.dna(sim_proteins,
               file = "sim_proteins.fasta",
               format = "fasta", 
               append = FALSE, 
               nbcol = 6,
               colsep = "", 
               colw = 10)

# Simulated protein alignment heatmap
readAAStringSet(filepath = "sim_proteins.fasta") %>%
  msa() %>%
  msaConvert(., type="seqinr::alignment") %>%
  dist.alignment(., "identity") %>%
  as.matrix(.) %>%
  heatmap()
```


```{r}
#mySequenceFile <- system.file("examples", file="lizard_proteins.fasta", package="msa")
#mySequences <- readAAStringSet(filepath = "lizard_proteins.fasta")
#alignment <- msa(mySequences)




#hemoAln2 <- msaConvert(alignment, type="seqinr::alignment")
#d <- dist.alignment(hemoAln2, "identity")
#as.matrix(d)[1:33, drop=FALSE]
#heatmap(as.matrix(d))
```


```{r}
# Write lizard proteins in a fasta
ape::write.dna(sim_proteins,
               file = "sim_proteins.fasta",
               format = "fasta", 
               append = FALSE, 
               nbcol = 6,
               colsep = " ", 
               colw = 10)

readAAStringSet(filepath = "sim_proteins.fasta") %>%
  msa() %>%
  msaConvert(., type="seqinr::alignment") %>%
  dist.alignment(., "identity") %>%
  as.matrix(.) %>%
  heatmap()
```


















