---
title: "lab 4"
author: "Simon Jorstedt"
date: "`r Sys.Date()`"
output:
  html_document:
    df_print: paged
---

# Question 1
*Run all the R code and reproduce the graphics. Go carefully through the R code and explain in your words what each step does. HINT Recall what a design/model matrix is from linear regression*

## Q 1.1: Working with Data - SummarisedExperiment

```{r Setup, message=FALSE, include=FALSE}
## ----style, echo = FALSE, results = 'asis'--------------------------------------------------------
BiocStyle::markdown()
options(width=100, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))

## ----setup, echo=FALSE, messages=FALSE, warnings=FALSE--------------------------------------------
suppressPackageStartupMessages({
    library(airway)
    library(DESeq2)
    library(ggplot2)
    library(org.Hs.eg.db)
})

```

This chunk gives an initial look at the dataset `airway`, which provides expression values in genes on four human airway muscle cell lines treated with a compound. We see that the data is organised as a matrix or assay with 63677 genes along the rows, and 8 different experiments along the columns. What follows below is a section of the assay,

```{r airway_experiment}
## ----airway-SummarizedExperiment------------------------------------------------------------------
library(airway)         # An 'ExperimentData' package...
data(airway)            # ...with a sample data set...
airway                  # ...that is a SummarizedExperiment
head(assay(airway))     # contains a matrix of counts
head(rowRanges(airway)) # information about the genes...
colData(airway)[, 1:3]  # ...and samples
## coordinated subsetting
untrt <- airway[, airway$dex == 'untrt']
head(assay(untrt))
colData(untrt)[, 1:3]

## ----airway-colData-------------------------------------------------------------------------------
library(airway)         # An 'ExperimentData' package...
data(airway)            # ...with a sample data set...
colData(airway)[, 1:3]  # ...represented as a SummarizedExperiment

## ----airway-assay---------------------------------------------------------------------------------
head(assay(airway))
```


Below, we perform a differential expression analysis, based on the Negative Binomial distribution, using the `DESeq2` package. This involves estimating size factors and dispersions, as well as fitting and testing the model. We then extract the results and order them from largest to smallest absolute log fold change. Below we present a table illustrating a subset of the results of the model. Among other things we report the p-value and log fold change for each gene.

```{r Analysis}
## ----airway-toptable------------------------------------------------------------------------------
library(DESeq2)     # package implementing statistical methods
dds <-              # data and experimental design
    DESeqDataSet(airway, design = ~ cell + dex)
dds <- DESeq(dds)   # initial analysis
res <- results(dds) # summary results
ridx <-             # order from largest to smallest absolute log fold change
    order(abs(res$log2FoldChange), decreasing=TRUE)
res <- res[ridx,]
head(res)           # top-table
```

Below we finally plot the negative logarithmic p-value against the log fold change. After that we obtain some information (gene symbol and name) for a small subset of the genes from the `org.Hs.eg.db` package.

```{r Visualization}
## ----airway-viz-----------------------------------------------------------------------------------
library(ggplot2)
g <- ggplot(as.data.frame(res), 
       aes(x=log2FoldChange, y=-10 * log10(pvalue))) +
    geom_point()
plot(g)
## ----airway-mapids--------------------------------------------------------------------------------
library(org.Hs.eg.db)
ensid <- head(rownames(res))
select(org.Hs.eg.db, ensid, c("SYMBOL", "GENENAME"), "ENSEMBL")
```

## Q 1.2: Genomic Ranges For Genome-Scale Data and Annotation

First we will investigate some properties of the IRanges package. We will create IRanges instances as a test. We also try out the flank method which returns left flanks of a certain length for given iranges.

```{r}
## ----iranges--------------------------------------------------------------------------------------
library(IRanges)
ir <- IRanges(start=c(10, 20, 30), width=5)
ir

## ----iranges-flank--------------------------------------------------------------------------------
flank(ir, 3)

## ----iranges-class--------------------------------------------------------------------------------
class(ir)
getClass(class(ir))

## ----iranges-flank-method, eval=FALSE-------------------------------------------------------------
#  ?"flank,Ranges-method"
```

Now we try out creating a test GRange, and find the left flanks of them.

```{r}
## ----granges--------------------------------------------------------------------------------------
library(GenomicRanges)
gr <- GRanges(c("chr1", "chr1", "chr2"), ir, strand=c("+", "-", "+"))
gr

## ----granges-flank--------------------------------------------------------------------------------
flank(gr, 3)

## ----granges-methods------------------------------------------------------------------------------
#methods(class="GRanges")

## ----granges-flank-method-------------------------------------------------------------------------
#grep("flank", methods(class="GRanges"), value=TRUE)

## ----granges-flank-method-help, eval=FALSE--------------------------------------------------------
#  ?"flank,GenomicRanges-method"

## ----granges-man-and-vignettes, eval=FALSE--------------------------------------------------------
#  help(package="GenomicRanges")
#  vignette(package="GenomicRanges")
#  vignette(package="GenomicRanges", "GenomicRangesHOWTOs")
```

Now below we load the `TxDb.Hsapiens.UCSC.hg19.knownGene` package, which contains a specific Human genome. Then we extract the genomic features from it using the `exons` function. Then we load another gene package, from which we extract the GC content of a particular sequence, which appears to be slightly larger than the average in the human genome.

```{r}
## ----txdb-----------------------------------------------------------------------------------------
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

## ----txdb-exons-----------------------------------------------------------------------------------
exons(txdb)

## ----txdb-exonsby---------------------------------------------------------------------------------
exonsBy(txdb, "tx")

## ----BSgenome-require, message=FALSE--------------------------------------------------------------
library(BSgenome.Hsapiens.UCSC.hg19)
chr14_range = GRanges("chr14", IRanges(1, seqlengths(Hsapiens)["chr14"]))
chr14_dna <- getSeq(Hsapiens, chr14_range)
letterFrequency(chr14_dna, "GC", as.prob=TRUE)

## ----bam-require----------------------------------------------------------------------------------
library(GenomicRanges)
library(GenomicAlignments)
library(Rsamtools)

## our 'region of interest'
roi <- GRanges("chr14", IRanges(19653773, width=1)) 
## sample data
library('RNAseqData.HNRNPC.bam.chr14')
bf <- BamFile(RNAseqData.HNRNPC.bam.chr14_BAMFILES[[1]], asMates=TRUE)
## alignments, junctions, overlapping our roi
paln <- readGAlignmentsList(bf)
j <- summarizeJunctions(paln, with.revmap=TRUE)
j_overlap <- j[j %over% roi]

## supporting reads
paln[j_overlap$revmap[[1]]]

## ----vcf, message=FALSE---------------------------------------------------------------------------
## input variants
library(VariantAnnotation)
fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
vcf <- readVcf(fl, "hg19")
seqlevels(vcf) <- "chr22"
## known gene model
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
coding <- locateVariants(rowRanges(vcf),
    TxDb.Hsapiens.UCSC.hg19.knownGene,
    CodingVariants())
head(coding)

## ----genomicalignments----------------------------------------------------------------------------
## example BAM data
library(RNAseqData.HNRNPC.bam.chr14)
## one BAM file
fl <- RNAseqData.HNRNPC.bam.chr14_BAMFILES[1]
## Let R know that this is a BAM file, not just a character vector
library(Rsamtools)
bfl <- BamFile(fl)

## ----readgalignments------------------------------------------------------------------------------
aln <- readGAlignments(bfl)
aln
```

## Q 1.3: Adding Annotation to your analysis

```{r}
## ----style, echo = FALSE, results = 'asis'--------------------------------------------------------
BiocStyle::markdown()
options(width=100, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))

## ----setup, echo=FALSE, messages=FALSE, warnings=FALSE--------------------------------------------
suppressPackageStartupMessages({
    library(AnnotationDbi)
    library(AnnotationHub)
    library(GenomicFeatures)
    library(biomaRt)
    library(org.Hs.eg.db)
    library(TxDb.Hsapiens.UCSC.hg19.knownGene)
})


## ----gene-model-discovery-------------------------------------------------------------------------
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene
txdb
methods(class=class(txdb))

## ----txdb-exons-----------------------------------------------------------------------------------
exons(txdb)
exonsBy(txdb, "tx")

## ----bsgenome-------------------------------------------------------------------------------------
library(BSgenome.Hsapiens.UCSC.hg19)
genome <- BSgenome.Hsapiens.UCSC.hg19
getSeq(genome, exons(txdb)[1:100])

## ----org------------------------------------------------------------------------------------------
library(org.Hs.eg.db)
org.Hs.eg.db

## ----select---------------------------------------------------------------------------------------
select(org.Hs.eg.db, c("BRCA1", "PTEN"), c("ENTREZID", "GENENAME"), "SYMBOL")
keytypes(org.Hs.eg.db)
columns(org.Hs.eg.db)

## ----organismdb-----------------------------------------------------------------------------------
library(Homo.sapiens)
select(Homo.sapiens, c("BRCA1", "PTEN"), 
       c("TXNAME", "TXCHROM", "TXSTART", "TXEND"), 
       "SYMBOL")

## ----biomart, eval=FALSE--------------------------------------------------------------------------
#  ## NEEDS INTERNET ACCESS !!
#  library(biomaRt)
#  head(listMarts(), 3)                      ## list marts
#  head(listDatasets(useMart("ensembl")), 3) ## mart datasets
#  ensembl <-                                ## fully specified mart
#      useMart("ensembl", dataset = "hsapiens_gene_ensembl")
#  
#  head(listFilters(ensembl), 3)             ## filters
#  myFilter <- "chromosome_name"
#  substr(filterOptions(myFilter, ensembl), 1, 50) ## return values
#  myValues <- c("21", "22")
#  head(listAttributes(ensembl), 3)          ## attributes
#  myAttributes <- c("ensembl_gene_id","chromosome_name")
#  
#  ## assemble and query the mart
#  res <- getBM(attributes =  myAttributes, filters =  myFilter,
#               values =  myValues, mart = ensembl)

## ----annotationhub-gtf, eval=FALSE----------------------------------------------------------------
#  library(AnnotationHub)
#  hub <- AnnotationHub()
#  hub
#  query(hub, c("Ensembl", "80", "gtf"))
#  ## ensgtf = display(hub)                   # visual choice
#  hub["AH47107"]
#  gtf <- hub[["AH47107"]]
#  gtf
#  txdb <- GenomicFeatures::makeTxDbFromGRanges(gtf)

## ----annotationhub-orgdb, eval=FALSE--------------------------------------------------------------
#  library(AnnotationHub)
#  hub <- AnnotationHub()
#  query(hub, "OrgDb")

## ----annotationhub-roadmap, eval=FALSE------------------------------------------------------------
#  library(AnnotationHub)
#  hub <- AnnotationHub()
#  query(hub , c("EpigenomeRoadMap", "E126", "H3K4ME2"))
#  E126 <- hub[["AH29817"]]

## ----annotationhub-liftover, eval=FALSE-----------------------------------------------------------
#  query(hub , c("hg19", "hg38", "chainfile"))
#  chain <- hub[["AH14150"]]

## ----liftover, eval=FALSE-------------------------------------------------------------------------
#  library(rtracklayer)
#  E126hg38 <- liftOver(E126, chain)
#  E126hg38

## ----vcf, message=FALSE---------------------------------------------------------------------------
## input variants
library(VariantAnnotation)
fl <- system.file("extdata", "chr22.vcf.gz", package="VariantAnnotation")
vcf <- readVcf(fl, "hg19")
seqlevels(vcf) <- "chr22"
## known gene model
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
coding <- locateVariants(rowRanges(vcf),
    TxDb.Hsapiens.UCSC.hg19.knownGene,
    CodingVariants())
head(coding)
```


## Q 1.4: Counting Reads and working with large files

```{r}
## ----style, echo = FALSE, results = 'asis'--------------------------------------------------------
BiocStyle::markdown()
options(width=100, max.print=1000)
knitr::opts_chunk$set(
    eval=as.logical(Sys.getenv("KNITR_EVAL", "TRUE")),
    cache=as.logical(Sys.getenv("KNITR_CACHE", "TRUE")))

## ----setup, echo=FALSE, messages=FALSE, warnings=FALSE--------------------------------------------
suppressPackageStartupMessages({
    library(GenomicFiles)
    library(BiocParallel)
})



## ----iteration------------------------------------------------------------------------------------
library(GenomicFiles)

yield <- function(bfl) {
    ## input a chunk of alignments
    library(GenomicAlignments)
    readGAlignments(bfl, param=ScanBamParam(what="seq"))
}

map <- function(aln) { 
    ## Count G or C nucleotides per read
    library(Biostrings)
    gc <- letterFrequency(mcols(aln)$seq, "GC")
    ## Summarize number of reads with 0, 1, ... G or C nucleotides
    tabulate(1 + gc, 73)                # max. read length: 72
}

reduce <- `+`

## ----iteration-doit-------------------------------------------------------------------------------
library(RNAseqData.HNRNPC.bam.chr14)
fls <- RNAseqData.HNRNPC.bam.chr14_BAMFILES
bf <- BamFile(fls[1], yieldSize=100000)
gc <- reduceByYield(bf, yield, map, reduce)
plot(gc, type="h",
     xlab="GC Content per Aligned Read", ylab="Number of Reads")

## ----parallel-doit--------------------------------------------------------------------------------
library(BiocParallel)
gc <- bplapply(BamFileList(fls), reduceByYield, yield, map, reduce)

library(ggplot2)
df <- stack(as.data.frame(lapply(gc, cumsum)))
df$GC <- 0:72
ggplot(df, aes(x=GC, y=values)) + geom_line(aes(colour=ind)) +
    xlab("Number of GC Nucleotides per Read") +
    ylab("Number of Reads")
```

## Q 1.5: RNA-Seq for understanding gene regulation

```{r}
## ----setup, echo=FALSE, messages=FALSE, warnings=FALSE--------------------------------------------
suppressPackageStartupMessages({
    library(DESeq2)
    library(limma)
    library(airway)
    library(gplots)
    library(RColorBrewer)
    library(ggplot2)
    library(genefilter)
    library(org.Hs.eg.db)
})

## -------------------------------------------------------------------------------------------------
library(airway)
data("airway")
se <- airway

## -------------------------------------------------------------------------------------------------
head(assay(se))

## -------------------------------------------------------------------------------------------------
colSums(assay(se))

## -------------------------------------------------------------------------------------------------
colData(se)

## ----rowRanges`-----------------------------------------------------------------------------------
rowRanges(se)

## -------------------------------------------------------------------------------------------------
library("DESeq2")
dds <- DESeqDataSet(se, design = ~ cell + dex)

## -------------------------------------------------------------------------------------------------
rld <- rlog(dds)
head(assay(rld))

## ----rldplot, fig.width=10, fig.height=5----------------------------------------------------------
opar <- par( mfrow = c( 1, 2 ) )
dds <- estimateSizeFactors(dds)
plot( log2( 1 + counts(dds, normalized=TRUE)[ , 1:2] ),
     col=rgb(0,0,0,.2), pch=16, cex=0.3 )
plot( assay(rld)[ , 1:2],
     col=rgb(0,0,0,.2), pch=16, cex=0.3 )
par(opar)

## -------------------------------------------------------------------------------------------------
sampleDists <- dist( t( assay(rld) ) )
sampleDists

## -------------------------------------------------------------------------------------------------
library("gplots")
library("RColorBrewer")

## ----distheatmap, fig.width=8---------------------------------------------------------------------
sampleDistMatrix <- as.matrix( sampleDists )
rownames(sampleDistMatrix) <- paste( rld$dex, rld$cell, sep="-" )
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
hc <- hclust(sampleDists)
heatmap.2( sampleDistMatrix, Rowv=as.dendrogram(hc),
          symm=TRUE, trace="none", col=colors,
          margins=c(2,10), labCol=FALSE )

## ----plotpca, fig.width=6, fig.height=4.5---------------------------------------------------------
plotPCA(rld, intgroup = c("dex", "cell"))

## -------------------------------------------------------------------------------------------------
(data <- plotPCA(rld, intgroup = c( "dex", "cell"), returnData=TRUE))
percentVar <- round(100 * attr(data, "percentVar"))

## -------------------------------------------------------------------------------------------------
library("ggplot2")

## ----ggplotpca, fig.width=6, fig.height=4.5-------------------------------------------------------
qplot(PC1, PC2, color=dex, shape=cell, data=data) +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance"))

## -------------------------------------------------------------------------------------------------
dds$dex <- relevel(dds$dex, "untrt")

## -------------------------------------------------------------------------------------------------
dds <- DESeq(dds)

## -------------------------------------------------------------------------------------------------
(res <- results(dds))

## -------------------------------------------------------------------------------------------------
mcols(res, use.names=TRUE)

## -------------------------------------------------------------------------------------------------
summary(res)

## -------------------------------------------------------------------------------------------------
results(dds, contrast=c("cell", "N061011", "N61311"))

## -------------------------------------------------------------------------------------------------
sum(res$pvalue < 0.05, na.rm=TRUE)
sum(!is.na(res$pvalue))

## -------------------------------------------------------------------------------------------------
sum(res$padj < 0.1, na.rm=TRUE)

## -------------------------------------------------------------------------------------------------
resSig <- subset(res, padj < 0.1)
head(resSig[ order( resSig$log2FoldChange ), ])

## -------------------------------------------------------------------------------------------------
head(resSig[ order( -resSig$log2FoldChange ), ])

## ----plotcounts, fig.width=5, fig.height=5--------------------------------------------------------
topGene <- rownames(res)[which.min(res$padj)]
data <- plotCounts(dds, gene=topGene, intgroup=c("dex"), returnData=TRUE)

## ----ggplotcountsdot, fig.height=5----------------------------------------------------------------
ggplot(data, aes(x=dex, y=count, fill=dex)) +
  scale_y_log10() + 
  geom_dotplot(binaxis="y", stackdir="center")

## ----plotma, eval=FALSE---------------------------------------------------------------------------
#  plotMA(res, ylim=c(-5,5))

## ----plotma2, eval=FALSE--------------------------------------------------------------------------
#  plotMA(res, ylim=c(-5,5))
#  with(res[topGene, ], {
#    points(baseMean, log2FoldChange, col="dodgerblue", cex=2, lwd=2)
#    text(baseMean, log2FoldChange, topGene, pos=2, col="dodgerblue")
#  })

## ----plotdispests---------------------------------------------------------------------------------
plotDispEsts(dds)

## ----histpvalue-----------------------------------------------------------------------------------
hist(res$pvalue, breaks=20, col="grey50", border="white")

## ----histpvalue2----------------------------------------------------------------------------------
hist(res$pvalue[res$baseMean > 1], breaks=20, col="grey50", border="white")

## -------------------------------------------------------------------------------------------------
library("genefilter")
topVarGenes <- head(order(-rowVars(assay(rld))),35)

## ----genescluster, fig.height=9-------------------------------------------------------------------
colors <- colorRampPalette( rev(brewer.pal(9, "PuOr")) )(255)
sidecols <- c("grey","dodgerblue")[ rld$dex ]
mat <- assay(rld)[ topVarGenes, ]
mat <- mat - rowMeans(mat)
colnames(mat) <- paste0(rld$dex,"-",rld$cell)
heatmap.2(mat, trace="none", col=colors, ColSideColors=sidecols,
          labRow=FALSE, mar=c(10,2), scale="row")

## ----sensitivityovermean, fig.height=4------------------------------------------------------------
# create bins using the quantile function
qs <- c(0, quantile(res$baseMean[res$baseMean > 0], 0:7/7))
# cut the genes into the bins
bins <- cut(res$baseMean, qs)
# rename the levels of the bins using the middle point
levels(bins) <- paste0("~",round(.5*qs[-1] + .5*qs[-length(qs)]))
# calculate the ratio of $p$ values less than .01 for each bin
ratios <- tapply(res$pvalue, bins, function(p) mean(p < .01, na.rm=TRUE))
# plot these ratios
barplot(ratios, xlab="mean normalized count", ylab="ratio of small p values")

## -------------------------------------------------------------------------------------------------
library(org.Hs.eg.db)

## -------------------------------------------------------------------------------------------------
columns(org.Hs.eg.db)
res$hgnc_symbol <- 
    unname(mapIds(org.Hs.eg.db, rownames(res), "SYMBOL", "ENSEMBL"))
res$entrezgene <- 
    unname(mapIds(org.Hs.eg.db, rownames(res), "ENTREZID", "ENSEMBL"))

## -------------------------------------------------------------------------------------------------
resOrdered <- res[order(res$pvalue),]
head(resOrdered)

## ----eval=FALSE-----------------------------------------------------------------------------------
#  write.csv(as.data.frame(resOrdered), file="results.csv")
```

## Q 1.6: ChIP-Seq for understanding gene regulation

```{r}
## ----setup, echo=FALSE, messages=FALSE, warnings=FALSE--------------------------------------------
suppressPackageStartupMessages({
    library(csaw)
    library(edgeR)
    library(GenomicRanges)
    library(ChIPseeker)
    library(genefilter)
    library(TxDb.Hsapiens.UCSC.hg19.knownGene)
    library(org.Hs.eg.db)
    library(clusterProfiler)
})

## ----null-p, cache=TRUE---------------------------------------------------------------------------
## 100,000 t-tests under the null, n = 6
n <- 6; m <- matrix(rnorm(n * 100000), ncol=n)
P <- genefilter::rowttests(m, factor(rep(1:2, each=3)))$p.value
quantile(P, c(.001, .01, .05))
hist(P, breaks=20)

## ---- eval=FALSE----------------------------------------------------------------------------------
#  vignette("ChIPseeker")

## ----csaw-setup-----------------------------------------------------------------------------------
files <- local({
    acc <- c(es_1="SRR074398", es_2="SRR074399", tn_1="SRR074417",
             tn_2="SRR074418")
    data.frame(Treatment=sub("_.*", "", names(acc)),
               Replicate=sub(".*_", "", names(acc)),
               sra=sprintf("%s.sra", acc),
               fastq=sprintf("%s.fastq.gz", acc),
               bam=sprintf("%s.fastq.gz.subread.BAM", acc),
               row.names=acc, stringsAsFactors=FALSE)
})


## ----load-csaw------------------------------------------------------------------------------------
data <- readRDS("csaw-data.Rds")

## ----csaw-filter----------------------------------------------------------------------------------
library(edgeR)     # for aveLogCPM()
keep <- aveLogCPM(assay(data)) >= -1
data <- data[keep,]

## ----csaw-normacs-load----------------------------------------------------------------------------
normfacs <- readRDS("csaw-normfacs.Rds")

## ----csaw-experimental-design---------------------------------------------------------------------
design <- model.matrix(~Treatment, colData(data))

## ----csaw-de--------------------------------------------------------------------------------------
y <- asDGEList(data, norm.factors=normfacs)
y <- estimateDisp(y, design)
fit <- glmQLFit(y, design, robust=TRUE)
results <- glmQLFTest(fit, contrast=c(0, 1))
head(results$table)

## ----csaw-merge-windows---------------------------------------------------------------------------
merged <- mergeWindows(rowRanges(data), tol=1000L)

## ----csaw-combine-merged-tests--------------------------------------------------------------------
tabcom <- combineTests(merged$id, results$table)
head(tabcom)

## ----csaw-grangeslist-----------------------------------------------------------------------------
gr <- rowRanges(data)
mcols(gr) <- as(results$table, "DataFrame")
grl <- split(gr, merged$id)
mcols(grl) <- as(tabcom, "DataFrame")
```



# Question 2
*In the presented analysis, there are no plots of raw paired data. In the section where the contrasts are defined and the three contrasts. Present the variables versus each other original, log scaled and MAplot for each considered pair both before and after normalization. A cluster analysis is performed on the page but not reported. Present plots and also draw heatmaps.*





